packages/components/src/data-model/word-query/word-query.js

```diff
@@ -0,0 +1,314 @@
+import { ClientAdapters } from 'alpheios-client-adapters'
+import WordQueryResult from '@comp/data-model/word-query/word-query-result.js'
+import WordQueryError from '@comp/data-model/word-query/word-query-error.js'
+
+/**
+ * Contains all the business logic that is necessary to resolve the GraphQL word query.
+ */
+export default class WordQuery {
+  constructor (options = {}, queryParams = {}) {
+    this._context = {
+      queryParams,
+      options: {
+        language: Language.fromJsonObject({ code: queryParams.language }),
+        clientID: options.clientID,
+        lexicons: options.lexicons,
+        shortLexicons: options.shortLexicons
+      },
+      cancelled: false,
+      result: undefined
+    }
+  }
+
+  start (result) {
+    this._context.result = result
+    this._context.result.loadingStarted()
+    if (this._context.queryParams.getLexemes) { this._context.result.state.lexemes.loading = true }
+    if (this._context.queryParams.getShortDefs) { this._context.result.state.shortDefs.loading = true }
+    const executionPath = WordQuery._createExecutionPath(this._context.queryParams)
+    try {
+      // Ignore result of an asynchronous operation and return immediately
+      // Result will be tracked via the proxied `result` variable
+      this._context = WordQuery._execute(executionPath, this._context)
+    } catch (err) {
+      Logger.getInstance().error(err)
+    }
+    return result
+  }
+
+  cancel () {
+    this._context.cancelled = true
+  }
+
+  // Creates a set of methods to be executed in order to fulfil the request.
+  static _createExecutionPath (queryParams) {
+    /*
+     Execution path consists of several batches. Batches are execution sequentially: each next batch
+     is executed after the previous one is completed.
+     Each batch consist of one or several jobs. Jobs within a batch can be executed in parallel.
+     */
+    let path = [] // eslint-disable-line prefer-const
+
+    let getLexemesBatch = [] // eslint-disable-line prefer-const
+    if (queryParams.getLexemes) {
+      if (queryParams.useMorphService) {
+        getLexemesBatch.push('_getTuftsMorphology')
+      }
+      if (queryParams.useTreebankData) {
+        getLexemesBatch.push('_getTreebankData')
+      }
+      if (queryParams.useWordAsLexeme) {
+        getLexemesBatch.push('_getWordAsLexeme')
+      }
+    }
+    if (getLexemesBatch.length > 0) {
+      path.push(getLexemesBatch)
+    }
+    path.push(['_disambiguate'])
+
+    if (queryParams.getShortDefs) {
+      path.push(['_getShortDefs'])
+    }
+    return path
+  }
+
+  // Runs all the jobs within the execution path.
+  static async _execute (executionPath, context) {
+    try {
+      for (const batch of executionPath) {
+        if (!context.cancelled) {
+          const jobs = batch.map(job => WordQuery[job](context))
+          await Promise.all(jobs)
```

Comment:
I didn't understand the construction here too.
what does this mean - `WordQuery[job](context)` without word `new`?

And also I didn't understand what is the heierarchy here and what would be do in sequence and what in parallel?
- job
- execution path
- batch

and when you pass `context` - to what object do you pass it - to the job or batch?
And if the batch is someting like - getMorph, disambiguate, getLexicons, the could contex be different for such steps? For example te result of previous step could add limits to the next steps?

Reply:
The job is a set of self sufficient operations such as retrieval of short definitions or of data from the Tufts adapter. Jobs are represented by their own static methods of the `WordQuery` class such as `_getTuftsMorphology()`, `_getTreebankData()`, `_disambiguate()`, `_getShortDefs()` and so on.

Jobs are grouped into batches. All jobs within one batch are executed in parallel.

A set of batches comprises the execution path. All batches are executed strictly one after the other. 

For example, if there is a word for which we need to retrieve data from the Tufts adapter and from the treebank, and get short definitions in addition to this then the execution path would be:
`[['_getTuftsMorphology', '_getTreebankData']['_disambiguate']['_getShortDefs']`.
It consists of three batches (top level array elements). The first batch consist of two jobs, the second and the third - of one. The content of the execution path arrays are the names of the methods of the `WordQuery` object.

The execution sequence for the execution path shown above would be:
1. `_getTuftsMorphology()` and `_getTreebankData()`, executed in parallel.
2. `_disambiguate()` - executed after the first two.
3. `_getShortDefs()` - executed the last.

`WordQuery[job]` refers to the static method of a WordQuery class (such as `WordQuery._disambiguate`). `WordQuery[job](context)` calls the method passing the `context` as a parameter. Because those methods are async, they return a promise. Those promises are gathered in the `jobs` array. `Promise.all(jobs)` converts multiple promises of a job into a single promise and in `await Promise.all(jobs)` we wait for all promises to resolve.

`context` is a shared state the is passed to each of the job methods. Each job (i.e. a static `WordQuery` method that matches the job) adds its own sub-object to the `context`. This sub-object contains the results of the job execution along with the status. For example, the `_getTuftsMorphology()` job adds the following to the context:
```
    context.tuftsMorphology = {
      state: {
        loading,
        available,
        failed
      },
      data: homonymGroup
    }
```
The `state` object contains results of the execution and the `data` prop contains the resulting `HomonymGroup` object.

Other jobs may use information from the previous jobs stored within the context. For example, a `_disambiguate` job may use Tufts and treebank adapter data from the context to produce a disambiguated homonym. It, for example, checks the `state.failed` field, and if it is truthy, then the results from the `data` field are not included into the disambiguation.

I will add a description of how it works to the documentation of the `WordQuery` class. Do you have any suggestions about how to make it clearer or how to simplify the workflow?

Reply:
I would think about the following structure (all titles are only as an example, and of course it is only my thoughts :) ):

- create a separate class jobOperation, that would have only static methods with all available jobOperations - `_getTuftsMorphology`, `_getTreebankData` and so on; and place here requirements of adding the current step - so we could see easily them. May be I would even create a separate folder with such classes , each file - is the one OperationStep, that has clear input, output, workflow and requirements in stadartized manner.

- create a separate class for ExecutionPath - and place there the hierarchy logic of combining steps, I would remove this complex if queue, that is not really readable and would be difficult to support; Here it would be good to place requirements to the order of steps, to the defining sequence or parallel they will be;

- and there will be only general methods in WordQuery class;

This way I would be able to separate layers of data and made them more independent, easier supported and easier expandable.

And this way I would have the following "bricks" for queries - JobOperation -> ExecutionPath -> Query

And also I didn't finally understand - would you have only one query - wordQuery that would retrieve all data, or would create several queries with different ExecutionsPaths inside?

If I would decide to have different Queries, for example GeneralLookupQuery, ShortDefsWordlistQuery, FullDataForWordItemQuery and so on, then I would create the following hierarchy:

 JobOperation -> ExecutionPath -> <specificQuery> -> Query
where Query would have general methods to work with AppoloClient, and <specificQuery> would have available executionPaths, requirements for input and output format for this query

@kirlat and @balmas , what do you think? Would such thoughts about code/data organization helpful?

Reply:
> I would think about the following structure

I like this suggestion, I was thinking about something along those lines. It also fits the DDD (Domain-Driven Design) very well. I will update the code accordingly.

Reply:
> And also I didn't finally understand - would you have only one query - wordQuery that would retrieve all data, or would create several queries with different ExecutionsPaths inside?

It is supposed to be one query, but with the query parameters that specify what data and from what sources should be retrieved.

Right now it uses the following parameters.
  `get` parameters specify what information should be returned:
    `getLexemes` - specifies that the lexemes need to be retrieved:
    `getShortDefs` - indicates that short definitions are needed;
  `use` parameters specify what sources should be used:
    `useMorphService` - specifies that a "main" morphology service should be used (such as Tufts);
    `useTreebankData` - specifies that data from treebank should be use;
    `useWordAsLexeme` - indicates that we should construct a lexeme out of a word;

Using different combination of parameters we can configure the query to return exactly what we need.

Does it makes sense? Are there any cases when this approach won't work?

Reply:
I think @irina060981 's suggestion for code organization is a good one.

Utlimately the challenge we are facing with this code is this:
1) we want to hide the complexity of how the different data sources are combined to make up the eventual word data that the user sees
2) at the same time, we want to make it possible for the user to have a say in what sources are used and how they are combined
3) and we want to be able to add to the available data sources as new resources become available

So we need to make the operations for retrieving the data as granular as possible, the algorithm for merging and disambiguating the data as flexible and self-describing as possible and the code for adding a new data source as simple as possible. It's not an easy challenge and I think it will take us several passes through to get it closer to where it needs to be.  

I believe that this is precisely the problem that the GraphQL architecture is designed to enable us to solve, but it will take some experience with it to figure out how to use it to full advantage.

Reply:
regarding the question of one or multiple queries, I believe the wordquery should be one query but we may have others for other use cases.  For now, I think retrieval of full definitions and other related resources do not warrant a graphql query. I think eventually the user wordlist will but not until we have other sorts of word-related user data.